src/metabase/models/action.clj

```diff
@@ -209,3 +214,160 @@
                         (m/index-by :id (actions-with-implicit-params (map :card dashcards) :id [:in action-ids])))]
     (for [dashcard dashcards]
       (m/assoc-some dashcard :action (get actions-by-id (:action_id dashcard))))))
+
+;;; ------------------------------------------------ Serialization ---------------------------------------------------
+
+(defmethod serdes.hash/identity-hash-fields Action
+  [_action]
+  [:name (serdes.hash/hydrated-hash :model "<none>") :created_at])
+
+(defmethod serdes.hash/identity-hash-fields QueryAction
+  [_action]
+  [(serdes.hash/hydrated-hash :action "<none>")])
+
+(defmethod serdes.hash/identity-hash-fields ImplicitAction
+  [_action]
+  [(serdes.hash/hydrated-hash :action "<none>")])
+
+(defmethod serdes.hash/identity-hash-fields HTTPAction
+  [_action]
+  [(serdes.hash/hydrated-hash :action "<none>")])
+
+(defmethod serdes.base/extract-one "Action"
+  [_model-name _opts action]
+  (-> (serdes.base/extract-one-basics "Action" action)
+      (update :creator_id serdes.util/export-user)
+      (update :model_id serdes.util/export-fk 'Card)))
+
+(defmethod serdes.base/load-xform "Action" [action]
+  (-> action
+      serdes.base/load-xform-basics
+      (update :model_id serdes.util/import-fk 'Card)
+      (update :creator_id serdes.util/import-fk-keyed 'User :email)))
+
+(defmethod serdes.base/extract-one "QueryAction"
+  [_model-name _opts query-action]
+  (-> (serdes.base/extract-one-basics "QueryAction" query-action)
+      (update :database_id serdes.util/export-fk-keyed 'Database :name)
+      (assoc :action_id (serdes.util/export-fk (:action_id query-action) 'Action))))
+
+(defmethod serdes.base/load-xform "QueryAction" [query-action]
+  (-> query-action
+      serdes.base/load-xform-basics
+      (update :database_id serdes.util/import-fk-keyed 'Database :name)
+      (update :action_id serdes.util/import-fk 'Action)))
+
+(defmethod serdes.base/extract-one "ImplicitAction"
+  [_model-name _opts implicit-action]
+  (-> (serdes.base/extract-one-basics "ImplicitAction" implicit-action)
+      (assoc :action_id (serdes.util/export-fk (:action_id implicit-action) 'Action))))
+
+(defmethod serdes.base/load-xform "ImplicitAction" [implicit-action]
+  (-> implicit-action
+      serdes.base/load-xform-basics
+      (update :action_id serdes.util/import-fk 'Action)))
+
+(defmethod serdes.base/extract-one "HTTPAction"
+  [_model-name _opts http-action]
+  (-> (serdes.base/extract-one-basics "HTTPAction" http-action)
+      (assoc :action_id (serdes.util/export-fk (:action_id http-action) 'Action))))
+
+(defmethod serdes.base/load-xform "HTTPAction" [http-action]
+  (-> http-action
+      serdes.base/load-xform-basics
+      (update :action_id serdes.util/import-fk 'Action)))
+
+(defmethod serdes.base/serdes-generate-path "Action"
+  [_ action]
+  (serdes.base/maybe-labeled "Action" action :name))
+
+(defmethod serdes.base/serdes-generate-path "QueryAction"
```

Comment:
We should discuss a redesign of the basic serdes setup for actions. Either the `QueryAction`/`ImplicitAction`/`HTTPAction` is inlined into the parent `Action`; or they are grouped together better.

I think I'm reading it right that each `Action` is exactly one of the three types; and that there is exactly one eg. `QueryAction` row for a given `Action`? In that case we almost certainly want to inline them into the Action.

If they're not 1-1, then we probably still want to nest the inner records of each `Action` under a common folder, perhaps something like:

```
- actions/
    - 12345678_some_action/
        - action.yml
        - query_action/
            - abcdef12_the_query_action.yml
            - 4321cafe_another_qa.yaml
```

Reply:
You're right, each `Action` is exactly one of three types, and they are 1:1 with their subtype model. I think inlining them into `Action` makes sense here.